{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eed501ab",
   "metadata": {},
   "source": [
    "# SCA\n",
    "### KO K00370 (nar)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b64301b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import to_hex\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import tqdm.notebook as tqdm\n",
    "from importlib import reload\n",
    "\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram, fcluster\n",
    "from scipy.spatial.distance import pdist\n",
    "\n",
    "import mysca.core as mysca_core\n",
    "import mysca.helpers as mysca_helpers\n",
    "from mysca.mappings import DEFAULT_MAP\n",
    "from mysca.io import load_msa\n",
    "from mysca.preprocess import preprocess_msa\n",
    "from mysca.core import run_sca, run_ica\n",
    "from mysca.helpers import get_top_k_conserved_retained_positions\n",
    "from mysca.helpers import get_conserved_rawseq_positions\n",
    "from mysca.helpers import get_rawseq_positions_in_groups\n",
    "from mysca.helpers import get_group_rawseq_positions_by_entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41cb158d",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEED = None\n",
    "rng = np.random.default_rng(seed=SEED)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa899663",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATDIRBASE = \"../data\"\n",
    "OUTDIRBASE = \"../out\"\n",
    "\n",
    "KOKEY = \"K00370\"\n",
    "# KOKEY = \"K00362\"\n",
    "\n",
    "SEQDIR = f\"{DATDIRBASE}/{KOKEY}/seqs\"\n",
    "MSADIR = f\"{DATDIRBASE}/{KOKEY}/msas\"\n",
    "STRUCTDIR = f\"{DATDIRBASE}/{KOKEY}/structures\"\n",
    "\n",
    "SEQ_FPATH = f\"{SEQDIR}/input.fasta\"\n",
    "MSA_FPATH = f\"{MSADIR}/MSA_800_with_reference.aln-fasta\"\n",
    "\n",
    "# REFSEQ_NAME = \"reference\"\n",
    "REFSEQ_NAME = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83be520d",
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTDIR = f\"{OUTDIRBASE}/sca_by_ko/{KOKEY}\"\n",
    "os.makedirs(OUTDIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7553505",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Convert letters to numbers (0-20 for gap + AAs)\n",
    "# AA_TO_INT = {aa: i for i, aa in enumerate(\"ACDEFGHIKLMNPQRSTVWY-\")}\n",
    "# GAP = AA_TO_INT[\"-\"]\n",
    "\n",
    "# AA_LIST = np.sort([k for k in AA_TO_INT.keys() if k != \"-\"])\n",
    "\n",
    "# NUM_AAS = len(AA_TO_INT)  # 20 amino acids + gap\n",
    "\n",
    "# assert NUM_AAS == 21, \"Should have 20 amino acids and 1 gap marker\"\n",
    "# assert GAP == 20, f\"GAP index should equal 20. Got {GAP}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bd65ffa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load MSA\n",
    "\n",
    "from mysca.mappings import SymMap\n",
    "DEFAULT_MAP = SymMap(\n",
    "    \"ACDEFGHIKLMNPQRSTVWY\", \"-\", [\"X\"]\n",
    ")\n",
    "\n",
    "msa_obj_orig, msa_orig, seqids_orig = load_msa(\n",
    "    MSA_FPATH, format=\"fasta\", \n",
    "    mapping=DEFAULT_MAP,\n",
    "    verbosity=1\n",
    ")\n",
    "\n",
    "NUM_SEQS, NUM_POS = msa_orig.shape\n",
    "print(f\"Loaded MSA shape: {msa_orig.shape} (sequences x positions)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e00d03a",
   "metadata": {},
   "outputs": [],
   "source": [
    "GAP_TRUNCATION_THRESH = 0.5\n",
    "SEQUENCE_GAP_THRESH = 0.5\n",
    "REFERENCE_ID = None\n",
    "REFERENCE_SIMILARITY_THRESH = 0.0\n",
    "SEQUENCE_SIMILARITY_THRESH = 1.0\n",
    "POSITION_GAP_THRESH = 0.25\n",
    "\n",
    "msa, xmsa, seqids, weights, fi0_pretrunc, retained_sequences, retained_positions, ref_results = preprocess_msa(\n",
    "    msa_orig, seqids_orig, \n",
    "    mapping=DEFAULT_MAP,\n",
    "    gap_truncation_thresh=GAP_TRUNCATION_THRESH,\n",
    "    sequence_gap_thresh=SEQUENCE_GAP_THRESH,\n",
    "    reference_id=REFERENCE_ID,\n",
    "    reference_similarity_thresh=REFERENCE_SIMILARITY_THRESH,\n",
    "    sequence_similarity_thresh=SEQUENCE_SIMILARITY_THRESH,\n",
    "    position_gap_thresh=POSITION_GAP_THRESH,\n",
    "    verbosity=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e4c2383",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "ax.plot(fi0_pretrunc, \".\")\n",
    "ax.hlines(POSITION_GAP_THRESH, *ax.get_xlim(), linestyle='--', color=\"r\", label=\"cutoff\")\n",
    "\n",
    "ax.legend()\n",
    "ax.set_xlim(0, 10 + msa.shape[1])\n",
    "\n",
    "ax.set_xlabel(f\"position\")\n",
    "ax.set_ylabel(f\"gap frequency\")\n",
    "ax.set_title(f\"Gap frequency by position\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb36c2b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "BACKGROUND_FREQ = {\n",
    "    'A': 0.078, 'C': 0.020, 'D': 0.053, 'E': 0.063,\n",
    "    'F': 0.039, 'G': 0.072, 'H': 0.023, 'I': 0.053,\n",
    "    'K': 0.059, 'L': 0.091, 'M': 0.022, 'N': 0.043,\n",
    "    'P': 0.052, 'Q': 0.042, 'R': 0.051, 'S': 0.071,\n",
    "    'T': 0.058, 'V': 0.066, 'W': 0.014, 'Y': 0.033\n",
    "}\n",
    "\n",
    "BACKGROUND_FREQ_ARRAY = np.zeros(20)\n",
    "for a in BACKGROUND_FREQ:\n",
    "    BACKGROUND_FREQ_ARRAY[DEFAULT_MAP[a]] = BACKGROUND_FREQ[a]    \n",
    "BACKGROUND_FREQ_ARRAY = BACKGROUND_FREQ_ARRAY / BACKGROUND_FREQ_ARRAY.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a8e4c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "LAM_REGULARIZATION = 0.03\n",
    "\n",
    "sca_results = run_sca(\n",
    "    xmsa, weights,\n",
    "    background_map=BACKGROUND_FREQ,\n",
    "    mapping=DEFAULT_MAP,\n",
    "    background_arr=BACKGROUND_FREQ_ARRAY,\n",
    "    regularization=LAM_REGULARIZATION,\n",
    "    return_keys=\"all\",\n",
    "    pbar=True,\n",
    "    leave_pbar=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5222a7c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fi0 = sca_results[\"fi0\"]\n",
    "fia = sca_results[\"fia\"]\n",
    "fijab = sca_results[\"fijab\"]\n",
    "Dia = sca_results[\"Dia\"]\n",
    "Di = sca_results[\"Di\"]\n",
    "Cijab_raw = sca_results[\"Cijab_raw\"]\n",
    "Cij_raw = sca_results[\"Cij_raw\"]\n",
    "phi_ia = sca_results[\"phi_ia\"]\n",
    "Cijab_corr = sca_results[\"Cijab_corr\"]\n",
    "Cij = sca_results[\"Cij_corr\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b52c1fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "N_TOP_CONSERVED = {\n",
    "    \"K00370\": 10,\n",
    "    \"K00362\": 5,\n",
    "}[KOKEY]\n",
    "topk_conserved_msa_pos, top_conserved_Di = get_top_k_conserved_retained_positions(\n",
    "    retained_positions, Di, N_TOP_CONSERVED\n",
    ")\n",
    "\n",
    "topk_conserved_msa_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d6c6dea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot conservation\n",
    "fig, ax = plt.subplots(1, 1, figsize=(10,4))\n",
    "\n",
    "ax.plot(\n",
    "    retained_positions, Di, \"o\",\n",
    "    color=\"Blue\",\n",
    "    alpha=0.2\n",
    ")\n",
    "\n",
    "ax.plot(\n",
    "    topk_conserved_msa_pos, top_conserved_Di, \"o\",\n",
    "    color=\"Green\",\n",
    "    alpha=0.5\n",
    ")\n",
    "\n",
    "ax.set_xlim(0, NUM_POS)\n",
    "ax.set_xlabel(f\"Position\")\n",
    "plt.ylabel(\"Relative Entropy (KL Divergence, $D_i$)\")\n",
    "ax.set_title(f\"Position-wise Conservation\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ab1cce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map MSA positions to raw sequence positions\n",
    "\n",
    "from mysca.helpers import get_rawseq_indices_of_msa\n",
    "\n",
    "rawseq_idxs = get_rawseq_indices_of_msa(msa_obj_orig)\n",
    "rawseq_idxs = rawseq_idxs[retained_sequences,:]\n",
    "rawseq_idxs = rawseq_idxs[:,retained_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fcc83e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Eigendecomposition of C_ij (raw and corrected)\n",
    "\n",
    "evals_sca_raw, evecs_sca_raw = np.linalg.eigh(Cij_raw)\n",
    "evals_sca_raw = np.flip(evals_sca_raw)\n",
    "evecs_sca_raw = np.flip(evecs_sca_raw, axis=1)\n",
    "\n",
    "evals_sca, evecs_sca = np.linalg.eigh(Cij)\n",
    "evals_sca = np.flip(evals_sca)\n",
    "evecs_sca = np.flip(evecs_sca, axis=1)\n",
    "\n",
    "print(f\"      Eigenvalue spectrum of Cij (raw): \" + \n",
    "      f\"{evals_sca_raw.min():.3g}, {evals_sca_raw.max():.3f}\")\n",
    "print(f\"Eigenvalue spectrum of Cij (corrected): \" + \n",
    "      f\"{evals_sca.min():.3g}, {evals_sca.max():.3f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "957c144b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "sc = ax.imshow(\n",
    "    Cij_raw, \n",
    "    cmap=\"Blues\", \n",
    "    origin=\"lower\",\n",
    "    vmax=None,\n",
    ")\n",
    "\n",
    "fig.colorbar(sc, label=\"Covariation\")\n",
    "ax.set_xlabel(\"(Retained) Position i\")\n",
    "ax.set_ylabel(\"(Retained) Position j\")\n",
    "ax.set_title(\"Covariance Matrix for K00370\")\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "sc = ax.imshow(\n",
    "    Cij, \n",
    "    cmap=\"Blues\", \n",
    "    origin=\"lower\",\n",
    "    vmax=None,\n",
    ")\n",
    "\n",
    "fig.colorbar(sc, label=\"Covariation\")\n",
    "ax.set_xlabel(\"(Retained) Position i\")\n",
    "ax.set_ylabel(\"(Retained) Position j\")\n",
    "ax.set_title(\"SCA Matrix for K00370\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4876c2b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Z = linkage(pdist(Cij, metric='euclidean'), method='ward')\n",
    "\n",
    "n_clusters = 10\n",
    "clusters = fcluster(Z, t=n_clusters, criterion='maxclust')\n",
    "\n",
    "\n",
    "dendro = dendrogram(Z, no_plot=True)\n",
    "leaf_indices = dendro['leaves']\n",
    "\n",
    "\n",
    "cmap = plt.cm.turbo\n",
    "cluster_colors = [to_hex(cmap(i)) for i in np.linspace(0, 1, n_clusters)]\n",
    "\n",
    "def color_func(link_idx):\n",
    "    if link_idx < len(clusters):  # Only color leaf nodes\n",
    "        return cluster_colors[clusters[link_idx] - 1]\n",
    "    return \"#000000\"\n",
    "\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(7, 6), \n",
    "                               gridspec_kw={'width_ratios': [0.2, 1]})\n",
    "\n",
    "\n",
    "dendrogram(\n",
    "    Z,\n",
    "    orientation='left',\n",
    "    ax=ax1,\n",
    "#    color_threshold=max(Z[-n_clusters+1, 2], 0.1),\n",
    "    link_color_func=color_func,\n",
    "    above_threshold_color='k'\n",
    ")\n",
    "\n",
    "ax1.set_ylabel('Position', fontsize='x-large')\n",
    "ax1.set_xticks([])\n",
    "ax1.set_yticks([])\n",
    "\n",
    "\n",
    "rearranged_data = Cij[leaf_indices][:, leaf_indices]\n",
    "im = ax2.imshow(\n",
    "    rearranged_data, \n",
    "    aspect='auto', \n",
    "    cmap='Blues',\n",
    "    interpolation='nearest', \n",
    "    origin='lower', \n",
    "    # vmin=0, vmax=1,\n",
    ")\n",
    "\n",
    "\n",
    "boundaries = np.where(np.diff(clusters[leaf_indices]))[0]\n",
    "for b in boundaries:\n",
    "    ax2.axhline(b + 0.5, color='black', linestyle='--')\n",
    "    ax2.axvline(b + 0.5, color='black', linestyle='--')\n",
    "\n",
    "ax2.set_title('Clustering of Positions', fontsize='x-large')\n",
    "ax2.set_xlabel('Position', fontsize='x-large')\n",
    "ax2.set_xticks([])\n",
    "ax2.set_yticks([])\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffca7562",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "fasta_fpath = SEQ_FPATH\n",
    "fastaseqs = SeqIO.parse(fasta_fpath, \"fasta\")\n",
    "fastaseqs = {str(e.id): e  for e in fastaseqs}\n",
    "fastaseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22cdc258",
   "metadata": {},
   "outputs": [],
   "source": [
    "conserved_aa_idxs = get_conserved_rawseq_positions(\n",
    "    msa_obj_orig, retained_sequences, topk_conserved_msa_pos\n",
    ")\n",
    "conserved_aa_idxs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cff009e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load PDB structures if available\n",
    "from mysca.io import get_residue_sequence_from_pdb_structure\n",
    "from mysca.io import load_pdb_structure\n",
    "\n",
    "pdb_mappings = {}\n",
    "missing_pdb_entries = []\n",
    "nan_filler = np.array([np.nan, np.nan, np.nan])\n",
    "for i, seqidx in enumerate(retained_sequences):\n",
    "    entry = msa_obj_orig[int(seqidx)]\n",
    "    id = entry.id\n",
    "    conserved_positions = conserved_aa_idxs[i]\n",
    "    pdbfpath = f\"{STRUCTDIR}/{id}.pdb\"\n",
    "    if not os.path.isfile(pdbfpath):\n",
    "        missing_pdb_entries.append(id)\n",
    "        continue\n",
    "    if -1 in conserved_positions:\n",
    "        print(f\"Entry {id} does not contain all conserved positions.\")\n",
    "        continue\n",
    "    structure = load_pdb_structure(pdbfpath, id=id, quiet=True)\n",
    "    residues = get_residue_sequence_from_pdb_structure(structure)\n",
    "    conserved_residues = [\n",
    "        residues[i] if i >= 0 else None for i in conserved_positions\n",
    "    ]\n",
    "    conserved_residue_positions = np.array(\n",
    "        [nan_filler if r is None else r['CA'].coord for r in conserved_residues]\n",
    "    )\n",
    "    pdb_mappings[id] = conserved_residue_positions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76aba3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97c240d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute pairwise distance matrix for conserved positions.\n",
    "\n",
    "ncombs = N_TOP_CONSERVED * (N_TOP_CONSERVED - 1) // 2\n",
    "all_pdists = np.nan * np.ones([len(pdb_mappings), ncombs])\n",
    "\n",
    "for i, id in enumerate(sorted(list(pdb_mappings.keys()))):\n",
    "    x = pdb_mappings[id]\n",
    "    dists = pdist(x, metric=\"euclidean\")\n",
    "    all_pdists[i] = dists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a9687f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "sc = ax.imshow(all_pdists, cmap = 'plasma')\n",
    "plt.ylabel('narG variant')\n",
    "plt.xlabel('pairwise distance')\n",
    "divider = make_axes_locatable(ax)\n",
    "cax = divider.append_axes('right', size='5%', pad=0.05)\n",
    "fig = ax.figure\n",
    "cbar = fig.colorbar(sc, cax=cax)\n",
    "cbar.ax.set_ylabel(\"Distance (Angstroms)\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5afd0f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "\n",
    "all_pdists_centered = (all_pdists - all_pdists.mean(0)) / all_pdists.std(0)\n",
    "\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "sc = ax.imshow(all_pdists_centered, cmap = 'plasma')\n",
    "plt.ylabel('narG variant')\n",
    "plt.xlabel('pairwise distance')\n",
    "divider = make_axes_locatable(ax)\n",
    "cax = divider.append_axes('right', size='5%', pad=0.05)\n",
    "fig = ax.figure\n",
    "cbar = fig.colorbar(sc, cax=cax)\n",
    "cbar.ax.set_ylabel(\"Distance (Normalized)\")\n",
    "\n",
    "pca = PCA(n_components=min(20, ncombs))\n",
    "pca.fit(all_pdists_centered)\n",
    "print(pca.explained_variance_ratio_)\n",
    "print(pca.singular_values_)\n",
    "data_pca = pca.transform(all_pdists_centered)\n",
    "\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "ax.plot(\n",
    "    data_pca[:,0], data_pca[:,1], \".\"\n",
    "\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(1, 1)\n",
    "ax.plot(\n",
    "    1 + np.arange(len(pca.explained_variance_ratio_)), \n",
    "    np.cumsum(pca.explained_variance_ratio_)\n",
    ")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "798ba0d4",
   "metadata": {},
   "source": [
    "## Bootstrapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0ecee5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "niters = 10\n",
    "\n",
    "def shuffle_columns(m, rng=None):\n",
    "    rng = np.random.default_rng(rng)\n",
    "    r, c = m.shape\n",
    "    idx = np.argsort(rng.random((r, c)), axis=0)\n",
    "    return m[idx, np.arange(c)]\n",
    "\n",
    "\n",
    "DO_SHUFFLING = True\n",
    "SHUFFLE_SEED = 13243\n",
    "shuffling_saveas = f\"{OUTDIR}/shuffled_cijs_corrected.npy\"\n",
    "\n",
    "if DO_SHUFFLING:\n",
    "    rng_shuffler = np.random.default_rng(SHUFFLE_SEED)\n",
    "    cijs_shuffled = np.full([niters, *Cij.shape], np.nan)\n",
    "    for iteridx in tqdm.trange(niters):\n",
    "        msa_shuff = shuffle_columns(msa, rng=rng_shuffler)\n",
    "        xmsa_shuff = np.eye(21, dtype=bool)[msa_shuff][:,:,:-1]\n",
    "        res = run_sca(\n",
    "            xmsa_shuff, weights,\n",
    "            background_map=BACKGROUND_FREQ,\n",
    "            mapping=DEFAULT_MAP,\n",
    "            background_arr=BACKGROUND_FREQ_ARRAY,\n",
    "            regularization=LAM_REGULARIZATION,\n",
    "            return_keys=[\"Cij_corr\"],\n",
    "            pbar=True,\n",
    "            leave_pbar=False,\n",
    "        )\n",
    "        cijs_shuffled[iteridx] = res[\"Cij_corr\"]\n",
    "\n",
    "    np.save(shuffling_saveas, cijs_shuffled)\n",
    "else:\n",
    "    cijs_shuffled = np.load(shuffling_saveas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0e07610",
   "metadata": {},
   "outputs": [],
   "source": [
    "evals_shuff = np.full([len(cijs_shuffled), *evals_sca.shape], np.nan)\n",
    "for i, cij_shuff in enumerate(cijs_shuffled):\n",
    "    evals = np.linalg.eigvalsh(cij_shuff)\n",
    "    evals_shuff[i] = np.flip(evals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b721d501",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "for e in evals_shuff:\n",
    "    ax.plot(\n",
    "        1 + np.arange(len(e)), e, \".\",\n",
    "        markersize=3\n",
    "    )\n",
    "\n",
    "\n",
    "ax.plot(\n",
    "    1 + np.arange(len(evals_sca)), evals_sca,\n",
    "    \"k.\",\n",
    "    markersize=2\n",
    ")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f586882",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "# Histogram of data eigenvalues\n",
    "counts, bins, patches = ax.hist(\n",
    "    evals_sca, bins=100, color=\"black\", alpha=0.8, log=True, label=\"Data\"\n",
    ")\n",
    "\n",
    "cutoff = np.mean(evals_shuff[:,1]) + 2 * np.std(evals_shuff[:,1])  # See SI G of [1]\n",
    "# cutoff = 100\n",
    "print(\"significant eigenvalue cutoff:\", cutoff)\n",
    "\n",
    "kstar = np.sum(evals_sca > cutoff)\n",
    "sig_evals_sca = evals_sca[:kstar]\n",
    "sig_evecs_sca = evecs_sca[:,:kstar]\n",
    "print(f\"{kstar} significant eigenvalues:\", sig_evals_sca)\n",
    "\n",
    "\n",
    "# Null distribution\n",
    "bin_centers = 0.5 * (bins[1:] + bins[:-1])\n",
    "h, bin_edges = np.histogram(evals_shuff.flatten(), bins=bins)\n",
    "ax.axvline(cutoff, 0, 1, linestyle=\"--\", color=\"grey\")\n",
    "ax.plot(bin_centers, h / niters, color=\"red\", lw=1.5, label=\"Null\")\n",
    "\n",
    "ax.legend()\n",
    "ax.set_xlabel(f\"$\\\\lambda$\")\n",
    "ax.set_ylabel(f\"Count\")\n",
    "ax.set_title(f\"Spectral decomposition\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0412599",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_evecs_sca.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b18e3b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "w_ica = run_ica(sig_evecs_sca.T)\n",
    "v_ica = sig_evecs_sca @ w_ica.T\n",
    "v_ica.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b58d82b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_ica_normalized = v_ica / np.sqrt(np.sum(np.square(v_ica), axis=0))\n",
    "for i in range(v_ica.shape[1]):\n",
    "    maxpos = np.argmax(np.abs(v_ica_normalized[:,i]))\n",
    "    if v_ica_normalized[maxpos,i] < 0:\n",
    "        v_ica_normalized[:,i] *= -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "539f58cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get groups\n",
    "groups = []\n",
    "for i in range(v_ica_normalized.shape[1]):\n",
    "    top95 = np.where(v_ica_normalized[:,i] >= np.percentile(v_ica_normalized[:,i], 95))[0]\n",
    "    groups.append(top95)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4064f3d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save groups in MSA coordinates\n",
    "\n",
    "subdir = f\"{OUTDIR}/groups\"\n",
    "os.makedirs(subdir, exist_ok=True)\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    np.save(f\"{subdir}/group_{i}_msapos.npy\", groups[i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8c0170",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "sc = ax.scatter(\n",
    "    v_ica_normalized[:,0], v_ica_normalized[:,1],\n",
    "    c='k', \n",
    "    # s=sizes, \n",
    "    # cmap='viridis', \n",
    "    alpha=0.2, \n",
    "    edgecolor='k'\n",
    ")\n",
    "\n",
    "for i, gidx in enumerate([0, 1]):\n",
    "    g = groups[gidx]\n",
    "    ax.scatter(\n",
    "        v_ica_normalized[g,0], v_ica_normalized[g,1],\n",
    "        # c=colors, \n",
    "        # s=sizes, \n",
    "        # cmap='viridis', \n",
    "        alpha=1, \n",
    "        edgecolor='k',\n",
    "        label=f\"Group {gidx}\"\n",
    "    )\n",
    "ax.legend(bbox_to_anchor=(1.05, 1), loc='upper left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437b82f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "\n",
    "sc = ax.scatter(\n",
    "    v_ica_normalized[:,0], v_ica_normalized[:,2],\n",
    "    c='k', \n",
    "    # s=sizes, \n",
    "    # cmap='viridis', \n",
    "    alpha=0.2, \n",
    "    edgecolor='k'\n",
    ")\n",
    "\n",
    "for i, gidx in enumerate([0, 1, 2, 3]):\n",
    "    g = groups[gidx]\n",
    "    ax.scatter(\n",
    "        v_ica_normalized[g,0], v_ica_normalized[g,2],\n",
    "        # c=colors, \n",
    "        # s=sizes, \n",
    "        # cmap='viridis', \n",
    "        alpha=1, \n",
    "        edgecolor='k',\n",
    "        label=f\"Group {gidx}\"\n",
    "    )\n",
    "ax.legend(bbox_to_anchor=(1.05, 1), loc='upper left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8a6ef53",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "fig = plt.figure(figsize=(12,5))\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "sc = ax.scatter(\n",
    "    v_ica_normalized[:,0], v_ica_normalized[:,1], v_ica_normalized[:,2], \n",
    "    c=\"k\", \n",
    "    # s=sizes, \n",
    "    # cmap='viridis', \n",
    "    alpha=0.2, \n",
    "    edgecolor='k'\n",
    ")\n",
    "\n",
    "for i, g in enumerate(groups):\n",
    "    ax.scatter(\n",
    "        v_ica_normalized[g,0], v_ica_normalized[g,1], v_ica_normalized[g,2], \n",
    "        # c=colors, \n",
    "        # s=sizes, \n",
    "        # cmap='viridis', \n",
    "        alpha=1, \n",
    "        edgecolor='k'\n",
    "    )\n",
    "\n",
    "# ax.scatter(0, 0, 0, \"r\")\n",
    "\n",
    "ax.set_xlabel(f\"IC 1\")\n",
    "ax.set_ylabel(f\"IC 2\")\n",
    "ax.set_zlabel(f\"IC 3\")\n",
    "ax.set_title(f\"ICA\")\n",
    "\n",
    "ax.view_init(elev=30, azim=40)   # elev ~ tilt, azim ~ around z; tweak to taste\n",
    "# ax.set_proj_type('persp')\n",
    "\n",
    "# plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3890dd51",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig = plt.figure(figsize=(12,5))\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "sc = ax.scatter(\n",
    "    v_ica_normalized[:,0], v_ica_normalized[:,1], v_ica_normalized[:,3], \n",
    "    c=\"k\", \n",
    "    # s=sizes, \n",
    "    # cmap='viridis', \n",
    "    alpha=0.2, \n",
    "    edgecolor='k'\n",
    ")\n",
    "\n",
    "for i, g in enumerate(groups):\n",
    "    ax.scatter(\n",
    "        v_ica_normalized[g,0], v_ica_normalized[g,1], v_ica_normalized[g,3], \n",
    "        # c=colors, \n",
    "        # s=sizes, \n",
    "        # cmap='viridis', \n",
    "        alpha=1, \n",
    "        edgecolor='k'\n",
    "    )\n",
    "\n",
    "# ax.scatter(0, 0, 0, \"r\")\n",
    "\n",
    "ax.set_xlabel(f\"IC 1\")\n",
    "ax.set_ylabel(f\"IC 2\")\n",
    "ax.set_zlabel(f\"IC 3\")\n",
    "ax.set_title(f\"ICA\")\n",
    "\n",
    "ax.view_init(elev=30, azim=40)   # elev ~ tilt, azim ~ around z; tweak to taste\n",
    "# ax.set_proj_type('persp')\n",
    "\n",
    "# plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c47e775",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save residue groups by sequence\n",
    "\n",
    "group_rawseq_positions = get_rawseq_positions_in_groups(\n",
    "    rawseq_idxs, groups\n",
    ")\n",
    "\n",
    "group_rawseq_positions_by_entry = get_group_rawseq_positions_by_entry(\n",
    "    msa_obj_orig, retained_sequences, groups, group_rawseq_positions\n",
    ")\n",
    "\n",
    "\n",
    "for groupidx in range(len(groups)):\n",
    "    subdir = f\"{OUTDIR}/sca_groups/group_{groupidx}\"\n",
    "    os.makedirs(subdir, exist_ok=True)\n",
    "    for i, seqidx in enumerate(retained_sequences):\n",
    "        entry = msa_obj_orig[int(seqidx)]\n",
    "        id = entry.id\n",
    "        pdbfpath = f\"{STRUCTDIR}/{id}.pdb\"\n",
    "        group_arr = group_rawseq_positions_by_entry[id][groupidx]\n",
    "        if os.path.isfile(pdbfpath):\n",
    "            np.save(f\"{subdir}/group_{groupidx}_{id}.npy\", group_arr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03bc3e2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_rawseq_positions_by_entry[msa_obj_orig[int(retained_sequences[0])].id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ace17e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !for f in data/K00370/structures/Soil*.pdb; do s=$(basename $f); s=${s/.pdb/}; echo $s; sh scripts/run_pymol_sca.sh $s; done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
